File: Day99/Part1/self_closing_dungeon.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <set>

using namespace std;

// Problem 1: Dungeon With Self-Closing Paths
// Input: N, M, Start, Target, InitialHP.
// Edges: u, v, min_hp, max_hp, delta_hp.
// Rule: Edge passed -> disappears. 
// State: (u, hp).
// Pruning: If we reach (u, hp) again, we have FEWER edges than before.
// Being at same state with fewer options is never better.
// So simple visited[(u, hp)] is valid for "Existence".

struct Edge {
    int v;
    int min_hp, max_hp;
    int delta_hp;
};

int main() {
    int N, M, Start, Target, InitHP;
    if (!(cin >> N >> M >> Start >> Target >> InitHP)) return 0;
    
    vector<vector<Edge>> adj(N + 1);
    for(int i=0; i<M; ++i) {
        int u, v, min_h, max_h, d_h;
        cin >> u >> v >> min_h >> max_h >> d_h;
        adj[u].push_back({v, min_h, max_h, d_h});
    }
    
    // BFS State: u, hp
    queue<pair<int, int>> q;
    set<pair<int, int>> visited;
    
    q.push({Start, InitHP});
    visited.insert({Start, InitHP});
    
    while(!q.empty()) {
        auto [u, hp] = q.front(); q.pop();
        
        if (u == Target) {
            cout << "YES" << endl;
            return 0;
        }
        
        for(auto& e : adj[u]) {
            if (hp >= e.min_hp && hp <= e.max_hp) {
                int next_hp = hp + e.delta_hp;
                if (next_hp > 0) { // Assume HP must be > 0 to survive
                    if (!visited.count({e.v, next_hp})) {
                        visited.insert({e.v, next_hp});
                        q.push({e.v, next_hp});
                    }
                }
            }
        }
    }
    
    cout << "NO" << endl;
    return 0;
}

File: Day99/Part1/self_closing_dungeon.py
import sys
from collections import deque

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    iterator = iter(input_data)
    try:
        N = int(next(iterator))
        M = int(next(iterator))
        Start = int(next(iterator))
        Target = int(next(iterator))
        InitHP = int(next(iterator))
        
        adj = [[] for _ in range(N + 1)]
        for _ in range(M):
            u = int(next(iterator))
            v = int(next(iterator))
            min_h = int(next(iterator))
            max_h = int(next(iterator))
            d_h = int(next(iterator))
            adj[u].append((v, min_h, max_h, d_h))
            
    except StopIteration:
        return

    # BFS
    # State: (u, hp)
    queue = deque([(Start, InitHP)])
    visited = set([(Start, InitHP)])
    
    while queue:
        u, hp = queue.popleft()
        
        if u == Target:
            print("YES")
            return
            
        for v, min_h, max_h, d_h in adj[u]:
            if min_h <= hp <= max_h:
                nxt_hp = hp + d_h
                if nxt_hp > 0:
                    if (v, nxt_hp) not in visited:
                        visited.add((v, nxt_hp))
                        queue.append((v, nxt_hp))
                        
    print("NO")

if __name__ == "__main__":
    solve()
